diff --git a/python/CMakeLists.txt b/python/CMakeLists.txt
index c544c490..efc081e7 100644
--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -11,7 +11,9 @@ set(AIE_PYTHON_ROOT_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
 # The AIE copy of the MLIR bindings is in the `aie.mlir` namespace.
 add_compile_definitions("MLIR_PYTHON_PACKAGE_PREFIX=aie.")
 # Upstream bindings emit due to `if (DerivedTy::getTypeIdFunction)`
-add_compile_options(-Wno-address)
+if(NOT WIN32)
+  add_compile_options(-Wno-address)
+endif()
 
 ################################################################################
 # Sources
