diff --git a/driver/src/common/xaie_txn.h b/driver/src/common/xaie_txn.h
index aa353c51..d06a5882 100644
--- a/driver/src/common/xaie_txn.h
+++ b/driver/src/common/xaie_txn.h
@@ -54,7 +54,8 @@ typedef enum {
 	XAIE_IO_CUSTOM_OP_MERGE_SYNC, // Previously this was XAIE_IO_CUSTOM_OP_BEGIN + 4
 	XAIE_IO_CUSTOM_OP_NEXT,
 	XAIE_IO_LOAD_PM_END_INTERNAL = 200,
-	XAIE_IO_CUSTOM_OP_MAX = UCHAR_MAX,
+	/* add new op here */
+	XAIE_IO_CUSTOM_OP_MAX
 } XAie_TxnOpcode;

 struct XAie_TxnCmd {
diff --git a/driver/src/core/xaie_core_aie.c b/driver/src/core/xaie_core_aie.c
index 188071fb..fdbc42ce 100644
--- a/driver/src/core/xaie_core_aie.c
+++ b/driver/src/core/xaie_core_aie.c
@@ -22,10 +22,6 @@
 *
 ******************************************************************************/
 /***************************** Include Files *********************************/
-#ifdef __linux__
-#include <unistd.h>
-#endif
-
 #include "xaie_core_aie.h"
 #include "xaie_feature_config.h"
 #include "xaie_helper_internal.h"
diff --git a/driver/src/global/xaiegbl.c b/driver/src/global/xaiegbl.c
index 83a1cebe..e55e7511 100644
--- a/driver/src/global/xaiegbl.c
+++ b/driver/src/global/xaiegbl.c
@@ -242,7 +242,7 @@ AieRC XAie_CfgInitialize(XAie_DevInst *InstPtr, XAie_Config *ConfigPtr)
 	memcpy(&InstPtr->PartProp, &ConfigPtr->PartProp,
 		sizeof(ConfigPtr->PartProp));

-	RC = XAie_IOInit(InstPtr);
+	RC = XAie_IOInit(InstPtr, ConfigPtr->Backend);
 	if(RC != XAIE_OK) {
 		return RC;
 	}
@@ -870,7 +870,7 @@ AieRC XAie_PerfUtilization(XAie_DevInst *DevInst, XAie_PerfInst *PerfInst)
 		XAIE_DBG("Start Col: %d\tnum: %d\n",
 				PartRange.Start, PartRange.Num);
 		PerfInst->Range = &PartRange;
-	} else if (PerfInst->Range->Num <= 0U ||
+	} else if (PerfInst->Range->Num == 0U ||
 			PerfInst->Range->Num > DevInst->NumCols) {
 		XAIE_ERROR("Invalid range!\n");
 		return XAIE_INVALID_ARGS;
diff --git a/driver/src/global/xaiegbl.h b/driver/src/global/xaiegbl.h
index 3529ea33..aad00440 100644
--- a/driver/src/global/xaiegbl.h
+++ b/driver/src/global/xaiegbl.h
@@ -187,6 +187,7 @@ typedef struct {
 	u8 AieTileRowStart;
 	u8 AieTileNumRows;
 	XAie_PartitionProp PartProp;
+	XAie_BackendType Backend;
 } XAie_Config;

 /*
diff --git a/driver/src/io_backend/ext/xaie_cdo.c b/driver/src/io_backend/ext/xaie_cdo.c
index 157ccbf8..c34c94d9 100644
--- a/driver/src/io_backend/ext/xaie_cdo.c
+++ b/driver/src/io_backend/ext/xaie_cdo.c
@@ -30,7 +30,13 @@

 #ifdef __AIECDO__ /* AIE simulator */

-#include "cdo_rts.h"
+#include "stdint.h"
+
+void cdo_Write32(uint64_t, uint32_t);
+void cdo_MaskWrite32(uint64_t , uint32_t, uint32_t);
+void cdo_MaskPoll(uint64_t , uint32_t, uint32_t, uint32_t);
+void cdo_BlockWrite32(uint64_t, const uint32_t*, uint32_t);
+void cdo_BlockSet32(uint64_t, uint32_t, uint32_t);

 #endif

diff --git a/driver/src/io_backend/ext/xaie_sim.c b/driver/src/io_backend/ext/xaie_sim.c
index 64391b93..a01db97e 100644
--- a/driver/src/io_backend/ext/xaie_sim.c
+++ b/driver/src/io_backend/ext/xaie_sim.c
@@ -35,7 +35,16 @@

 #ifdef __AIESIM__ /* AIE simulator */

-#include "main_rts.h"
+#include <stdint.h>
+
+typedef unsigned int uint;
+
+void ess_Write32(uint64_t Addr, uint Data);
+uint ess_Read32(uint64_t Addr);
+void ess_WriteCmd(unsigned char Command, unsigned char ColId, unsigned char RowId, unsigned int CmdWd0, unsigned int CmdWd1, unsigned char *CmdStr);
+
+void ess_NpiWrite32(uint64_t Addr, uint Data);
+uint ess_NpiRead32(uint64_t Addr);

 #endif

diff --git a/driver/src/io_backend/xaie_io.c b/driver/src/io_backend/xaie_io.c
index b3aa3753..247e17aa 100644
--- a/driver/src/io_backend/xaie_io.c
+++ b/driver/src/io_backend/xaie_io.c
@@ -121,10 +121,17 @@ static const XAie_Backend *IOBackend[XAIE_IO_BACKEND_MAX] =
 * @note		Internal Only.
 *
 ******************************************************************************/
-AieRC XAie_IOInit(XAie_DevInst *DevInst)
+AieRC XAie_IOInit(XAie_DevInst *DevInst, XAie_BackendType backend)
 {
 	AieRC RC;
-	const XAie_Backend *Backend = IOBackend[XAIE_DEFAULT_BACKEND];
+	const XAie_Backend *Backend;
+
+	if (backend >= XAIE_IO_BACKEND_MAX) {
+		XAIE_DBG("Invalid backend %d; using default\n", backend);
+		backend = XAIE_DEFAULT_BACKEND;
+	}
+
+	Backend = IOBackend[backend];

 	RC = Backend->Ops.Init(DevInst);
 	if(RC != XAIE_OK) {
diff --git a/driver/src/io_backend/xaie_io.h b/driver/src/io_backend/xaie_io.h
index f2795377..d667274e 100644
--- a/driver/src/io_backend/xaie_io.h
+++ b/driver/src/io_backend/xaie_io.h
@@ -174,7 +174,7 @@ struct XAie_Backend {
 };

 /************************** Function Prototypes  *****************************/
-AieRC XAie_IOInit(XAie_DevInst *DevInst);
+AieRC XAie_IOInit(XAie_DevInst *DevInst, XAie_BackendType backend);

 #endif	/* End of protection macro */

